algorithm to read psd event data for the DNS instrument at MLZ

thomueller opened the pull request:
Fixes #22871.

The PR adds the algorithm LoadDNSEvent, which is a loader for the Event data of the position sensitive detector of the DNS instrument at MLZ. It loads the events into an EventWorkspace.

Requested by MLZ DNS instrumenent scientist = me.

**To test:**
This is also the example in the docs:

    eventWS = LoadDNSEvent(InputFile="DNS_psd_150c_first_tube.mdat", ChopperChannel='2', NumberOfTubes='1')
    print("Number of events: {}".format(eventWS.getNumberEvents()))
    print("Maximum time of flight: {}".format(eventWS.getTofMax()))
    print("Number of detector pixels: {}".format(eventWS.getNumberHistograms()))
    print("Number of bins: {}".format(eventWS.blocksize()))
    # rebin spectra:
    test2 = Rebin(InputWorkspace='eventWS', Params='0, 100, {0}'.format(eventWS.getTofMax()))
    print("Number of bins after rebinning: {}".format(test2.blocksize()))

Should output:
    Number of events: 184
    Maximum time of flight: 6641.4
    Number of detector pixels: 1024
    Number of bins: 1
    Number of bins after rebinning: 67

For release notes see [here](https://github.com/thomueller/mantid/commit/873d8f09e617a205fac722458254c980741c95e9).


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

thomueller committed:
first_working_version_psd_reader

thomueller committed:
move h files

thomueller committed:
Merge branch 'mantidproject:master' into dns_psd_reader

thomueller committed:
first working version

thomueller committed:
Merge branch 'dns_psd_reader' of github.com:thomueller/mantid into dns_psd_reader

thomueller committed:
first working version, with working tests, checked against nicos

thomueller committed:
changed example

thomueller committed:
filename changed

thomueller committed:
files added to repo

thomueller committed:
files added to repo

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
add release notes

thomueller committed:
fix format to clang

thomueller committed:
fix format to clang

thomueller committed:
remove blanks

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
update example

thomueller committed:
added empty constructor in h file

thomueller committed:
change algm registration to fielloader

thomueller committed:
remove if check if alreday defined

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
add alg to cmake file

thomueller committed:
fix cppcheck errors

thomueller committed:
ignore cppcheck warning on ending change

thomueller committed:
fix example

thomueller committed:
fix example

thomueller committed:
fix_unit_test

thomueller committed:
increase thread count to 8

darominski committed:
Added autocompletion switch to settings UI.

darominski committed:
Connected autocompletion switch with code editor

thomueller committed:
add pregenerated vtp file for testing

thomueller committed:
add pregenerated vtp file for testing

thomueller committed:
add unittest for confidence and fix preloading vtp

thomueller committed:
move vtp loading to setup

thomueller committed:
remove most tests, to find error

thomueller committed:
add first execution test

thomueller committed:
fix override warning

thomueller committed:
fix override warning2

thomueller committed:
remove all test but one, and explicitly load vtp

thomueller committed:
change vtkdirectroykey

thomueller committed:
change vtk directtory key back

thomueller committed:
remove threading

thomueller committed:
resolve release note conflict

thomueller committed:
resolve release note conflict

thomueller committed:
Merge pull request #1 from mantidproject/main

merge mantid main

thomueller committed:
add all tests

thomueller commented:
Hi David, 
thanks for you help. I was not award of the special section on the forum.
I think I fixed the merge conflict now, was only the documentation change. 
Now it has some error in the merge conflict check, probably I did something wrong, while merging the changes form mantid main.
Error: GraphqlError: Resource not accessible by integration

It seems to compile well also on OSX, the problem is running the tests, running the test takes too long on OSX.
On Windows the test runs just fine.
Previously there was a message that it was creating the vtp file, so I thought this was the problem.
And  DanielMurphy22, pointed out that one can precreate the file, which I did, but this seems not to fix the problem.
https://forum.mantidproject.org/t/my-osx-build-fails-probably-because-the-geometry-cache-building-takes-too-long/789/2

thomueller committed:
fix test

thomueller commented:
So the problem seems to be that, the Action now tries to change labels on the mantid main, which it should not do and luckily cannot do.

thomueller commented:
Ok now it seems fine, just with the Error on OSX.
My assumption is that is has something to do with the byteorder, in the BitStream.h there are some conversion for different byteorders, written beim Joachim Coenen. But I have no OSX system to test them.

thomueller committed:
change IDF, change test to load chopper channel from IDF

thomueller committed:
update test to load new vtp file, new md5 hash

DavidFair commented:
So I've cleared the instrument cache locally and this test still ran.
I'm concerned about uploading a pre-generated instrument cache as it can easily break in the future and my gut feeling is this will still fail on OSX CI

If it does fail again I'll SSH and run through it on LLDB as I worry we're deadlocking somewhere

thomueller commented:
Hi David,
it has also failed with the pregenerated instrument file before. 
I also do not think it is the problem. 
The loading from the cache message was showing up in the failed test log, so I thought that was the problem, but apparently it is not.

DavidFair commented:
When the user workshop finishes I'll offline the builder and have a go with the debugger to figure out where it's getting stuck. This could be something related to XCode or some OSX specific code deep in Mantid

Could you pull out the changes bringing in the cached file, since we've eliminated that and it saves noise whilst I'm debugging?
Regarding the test size, is there a way to create a subset of the data, such as only saving out the first 10 spectra for example? 

This helps reduce the test times down in the future, especially if we assume loading the first n spectra is a good indicator of the remaining spectra

thomueller committed:
remove pregenrated vtp

thomueller commented:
Hi David,
i have delete the pregenerated vtp part.
Regarding reducing the spectra I am not sure how to realize this. 
Since the spectra are the detector pixels.
Only way I can see is create a reduced instrument. Or change the algorithm to create only filled spectra, but I am not sure how this will work then with further data reduction like instrument view.
Changing the datafile will not reduce the spectra.
I can further reduce the number of events, is about 10k atm., thought that was small, but probably I can also live with 1000.
Although it seems that loading the data itself is not the speed critical factor, it is rather the creation of the algorithm for each test. I had some plans to improve that, by merging some test.

thomueller commented:
Atm. the algorithm creates an Eventworkspace with a fixed size of 1024x128 spectra.
Which is not smart anyhow since we only have 1024x64 detectors installed atm.
Made some quick test, this is the slowing down the algorithm a lot (as you suggested reducing the spectra is key).
I will change this, but I have to think about it how to do it.
I need to create some real data, distributed on less detectors.
And then load a reduced instrument for testing.
Problem is, that the algorithm, also calculates the indexing of the detectors, which I want to check, since it was done wrong before. Probably have to just check the indexing function just separately, and not by loading real data through the algorithm.

DavidFair commented:
So this is something screwy happening with the build on our CI.

It's the geometry cache definitely, locally it takes <1s
```
InstrumentDefinitionParser-[Notice] Geometry cache is not available
InstrumentDefinitionParser-[Notice] Creating cache in ../DNS-PSDbfbdb95375d9e6c728c1b2439e7c9c443060349c.vtp
LoadDNSEvent-[Notice] LoadDNSEvent successful, Duration 0.89 seconds
```

On the remote CI it took 2 minutes when I finally cancelled out. Let me attach a profiler and see if it's broken codegen

DavidFair commented:
I'm not in tomorrow (Wednesday) but it looks like it's going to take some co-ordination on our end as I think it's a compiler / libc++ bug based on the fact:
- It works fine with XCode 12
- Every other platform works find
- The stack trace looks like a hang we had with the address sanitizers, when they weren't playing nice with exception in libc++

However, we will need to upgrade the OS on the builder to know for sure...

thomueller commented:
Hi David,
thanks for all the work you put into it. This sounds like something I would not have been able to figure out alone.

thomueller committed:
speed up unittests

thomueller committed:
twim whitespace in rst

thomueller committed:
fix cppcheck

thomueller committed:
add md5 for testing

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
remove instrument loading, fix docs example, change to tubes as input parameter

thomueller commented:
I removed the instrument loading from the algorithm. It was not used for anything. Beside adding the instrument to the workspace at the end. 
Removing it makes the algorithm faster, avoids the OSX problem and makes the algorithm usable for other instruments which use Mesytec even listmode datafiles.

If the instrument definition is needed one can simply run LoadInstrument on the workspace afterwards.

thomueller commented:
So OSX build still fails, with LoadInstrument completly removed, seems to have nothing to do with the geometry cache.

thomueller committed:
add end of file break condition in split function

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller commented:
Ok seems I found the problem. On OSX the splitting of the file goes into an endless loop, since there was no check for EOF, but still not sure why this only affects OSX.

thomueller committed:
Merge branch 'main' into 22871_read_DNS_psd_event_data

thomueller committed:
fix warning of ref link in docs

DavidFair commented:
So @martyngigg looked at this on OSX last week (sorry I was on leave Monday so you beat me to it figuring out the problem),

It appears on older versions of LLVM (and Apple Clang) there is some ambiguity on when an exception is thrown reading from a file. For MSVC and GCC it will throw trying to read from a handle which has reached the EOF, unless you opt into the exception explicitly (see https://stackoverflow.com/a/15667188).

In recent versions of LLVM (the last 1/2  years) this has moved more in-line with GCC and MSVC, hence we couldn't reproduce it. Additionally, most of Mantid will check for EOF in a loop rather than relying on exception control flow.
It appeared to me like a deadlock at the OS level, as effectively libc++ would throw then internally catch the exception, and we would continue to try parsing the same line again in a tight loop.

If you want to restore OpenMP / any of the other optimisations you took out through the course of debugging I'd be happy to have them again now we understand the underlying problem

thomashampson commented:
I've moved this to the 6.4 milestone, but if you want it to be in 6.3 you'll have to change the milestone back and target release-next rather than main.

DavidFair commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
This file has been added by accident, can we remove it

thomueller replied:
ups, removed

DavidFair commented on docs/source/algorithms/LoadDNSEvent-v1.rst:
```diff
@@ -0,0 +1,78 @@
+    :ref:`EventWorkspace <EventWorkspace>` with the X-axis in the Time-of-flight units.
+
+**ChopperChannel**
+is the input channel to wich the chopper was connected, must also be set in mesydaq gui.
```
Could we rephrase this to:
`The chopper input channel, must also...`

thomueller replied:
sure, changed it to: The chopper input channel as set in mesydaq gui.

DavidFair reviewed (changes requested):
My priority is to get this into 6.3 if you want it to, as it's been open for a long while. Sorry about the delay reviewing I was on leave for a period of time.

I've checked and we can still work on this past the feature freeze that was the 24th (as it's a completely new isolated feature), but we should be aiming to wrap it up during this week to make it in time.

The requested changes boils down to:
- Use boost / STL functionality where available in a number of places instead of rolling our own
- Some initialization safety improvements
- Some coding standards tweaks
- Some readability fixes

This is a partial review currently, I suspect there will be more changes to `BitStream` as a lot of the manual endian handling will go away making it easier to re-review

@thomueller Will you have the dev time available to work through the changes in the coming week or would you like to push it back to 6.4?
I'm happy either way.

darominski committed:
Merge branch 'main' into 32862_codeeditor_autocompletion_switch

darominski committed:
Release note for code editor auto-completion setting

darominski committed:
CodeEditor autocompletion setting unit tests

sf1919 committed:
Updated release script to fix errors

sf1919 committed:
Skeleton release notes for v6.4

martyngigg committed:
Merge pull request #33210 from mantidproject/6.4_skeleton_release_notes

Create v6.4 skeleton release notes

Harrietbrown committed:
rename single fitting button in IDA

This commit changes the wording for the button to perform a single fit in indirect data analysis.

Commit:
Merge remote-tracking branch 'origin/ornl-next' into main

thomueller commented:
Hi David, 
I was on vacation and read your message only today.
I will work in it today, but I guess that's to late.
Thanks for you review. This are all good points, especially the replacement with the functions from boots library.

Commit:
Merge pull request #33242 from mantidproject/rename_fit_single_button

rename single fitting button in IDA

darominski committed:
Merge remote-tracking branch 'origin/main' into 32862_codeeditor_autocompletion_switch

darominski committed:
Moved the auto-completion switch release note to v6.4

martyngigg committed:
Merge pull request #32933 from mantidproject/32862_codeeditor_autocompletion_switch

CodeEditor auto-completion switch

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  }
+}
+
+namespace {
+template <typename Iterable>
+constexpr std::map<std::remove_reference_t<decltype(*std::declval<decltype(std::declval<Iterable>().cbegin())>())>,
+                   size_t>
+buildSkipTable(const Iterable &iterable) {
+  std::map<std::remove_reference_t<decltype(*std::declval<decltype(std::declval<Iterable>().cbegin())>())>, size_t>
+      skipTable;
+
+  size_t i = iterable.size();
+  for (auto c : iterable) {
+    i--;
+    auto &cPos = skipTable[c];
+    cPos = cPos == 0 ? iterable.size() : cPos;
+    cPos = i == 0 ? static_cast<size_t>(cPos) : i;
+  }
+  return skipTable;
+}
+
+template <typename Iterable> const std::vector<uint8_t> buildSkipTable2(const Iterable &iterable) {
+  std::vector<uint8_t> skipTable(256, static_cast<uint8_t>(iterable.size()));
+
+  size_t i = iterable.size();
+  for (auto c : iterable) {
+    i--;
+    auto &cPos = skipTable[c];
+    cPos = i == 0 ? cPos : static_cast<uint8_t>(i);
+  }
+  return skipTable;
+}
```
I am not sure, how to replace this. The code is not simply doing string search.  It is incrementally loading the data file, in the steps given by Boyer Moore. If I want to do pure string search, I would have to load an unknown large part of the datafile.

thomueller replied:
Ah I think we can use it at least to create the table.

DavidFair replied:
Have we measured the time in doing this search on-disk vs loading all data in then parsing / filtering in-memory?
I worry this is a case of premature optimisation, if it's a significant difference I'm happy to take it. But I worry we're trading a couple of seconds of loading time (max) for significantly harder to maintain code

thomueller replied:
Hi David,
I have not originally written this, so I have not done any test if it was necessary. Was originally written by Joachim Coenen. I just fixed everything which was broken.

Loading everything in memory is not really an option I think, we are talking about hunderts of Mb per data file file and Gbs in total.
I would rather try, to do bytewise incrementally load, and check every time if the string matches, with boost or normal string  search. This would only be a marginal change to the code and would increase the checks by the time you would save by Boyer Moore, but could well be irrelevant for the total loading time. 
String matching should be very fast. For the header, it is only 2 lines to check at DNS, so the speed difference should be marginal. For the blocks its 1500byte, and around 8 blocks for file splitting. 
Compared to Billions of events which are processed, this should not matter.

Just made some quick tests, setting the skip table to 1 byte for everything.
There is no visible difference in speed, it is hidden in the fluctuations, for 1 Billion of counts less than 0.02 s.

DavidFair replied:
With modern optimising compilers I wouldn't be surprised if there's some tricks being inlined that causes the naive approach to be pretty speedy.

Thanks for testing that, I was on the fence about asking since you've already done lots of work for this PR. But good to see it all ultimately worked out :+1:

thomueller committed:
fix everythin David suggested, beside the Boyer Moore search

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/BitStream.h:
```diff
@@ -0,0 +1,230 @@
+#include <sys/stat.h>
+#include <vector>
+
+/**
```
changed

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/BitStream.h:
```diff
@@ -0,0 +1,230 @@
+    data[i] = data[i] ^ data[j];
+  }
+  return value;
+}
```
dropped completely with conditional reverse from boost

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/BitStream.h:
```diff
@@ -0,0 +1,230 @@
+  const endian machineEndianess = (arrayValue.asUint == mixInt) ? endian::big : endian::little;
+  return machineEndianess;
+}
+} // namespace
```
dropped completely with conditional reverse from boost

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/BitStream.h:
```diff
@@ -0,0 +1,230 @@
+  // template <typename T> inline T read(T &&result) {
+  // read<sizeof(result)>(result);
+  // return result;
+  // }
```
removed

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+#include <array>
+#include <fstream>
+#include <limits>
+//#include <iterator>
```
removed it also the rest, since it was unnecessary includes

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+
+namespace Mantid::DataHandling {
+
+/**
```
fixed

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+  bool discardPreChopperEvents;
+  bool setBinBoundary;
+
+  void populate_EventWorkspace(Mantid::DataObjects::EventWorkspace_sptr eventWS,
```
passed by reference now

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+    uint16_t runId;
+    uint8_t mcpdId;
+    uint8_t deviceStatus;
+    uint64_t timestamp;
```
all set to 0 at initialization

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+
+  // uint32_t monitorChannel;
+  bool discardPreChopperEvents;
+  bool setBinBoundary;
```
also set to default values

thomueller commented on Framework/DataHandling/inc/MantidDataHandling/LoadDNSEvent.h:
```diff
@@ -0,0 +1,126 @@
+  };
+
+  uint32_t chopperChannel;
+  uint32_t detectorPixelCount;
```
I think I added m_ now to all member variables.

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+    result[j + 1] = digits[wtmp_p[i] & 0x0F];
+  }
+  return result;
+}
```
dropped completely with conditional reverse from boost

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+}
+
+const unsigned MAX_BUFFER_BYTES_SIZE = 1500; // maximum buffer size in data file
+const unsigned PIXEL_PER_TUBE = 1024;        // maximum buffer size in data file
```
sure, done

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  const std::string fileName = getPropertyValue("InputFile");
+  chopperChannel = static_cast<uint32_t>(getProperty("ChopperChannel"));
+  discardPreChopperEvents = static_cast<bool>(getProperty("DiscardPreChopperEvents"));
+  setBinBoundary = static_cast<bool>(getProperty("SetBinBoundary"));
```
ok done

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+
+template <typename Vector, typename _Compare> void sortVector(Vector &v, _Compare comp) {
+  std::sort(v.begin(), v.end(), comp);
+}
```
true, done

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+    // PARALLEL_END_INTERUPT_REGION
+    oversizedChanelIndexCounterA += oversizedChanelIndexCounter;
+    oversizedPosCounterA += oversizedPosCounter;
+    // PARALLEL_CHECK_INTERUPT_REGION
```
added back

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  sortVector(finalEventAccumulator.triggerEvents, [](auto l, auto r) { return l.timestamp > r.timestamp; });
+
+  std::atomic<uint64_t> oversizedChanelIndexCounterA(0);
+  std::atomic<uint64_t> oversizedPosCounterA(0);
```
yes should be fine, should not get larger than 31 * 1024 * 128, changed

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  std::atomic<uint64_t> oversizedPosCounterA(0);
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*eventWS) && USE_PARALLELISM)
+  for (int j = 0; j < static_cast<int>(finalEventAccumulator.neutronEvents.size()); ++j) {
```
renamed

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+    // uint64_t chopperTimestamp = 0;
+    uint64_t oversizedChanelIndexCounter = 0;
+    uint64_t oversizedPosCounter = 0;
+    uint64_t i = 0;
```
renamed and moved

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+    }
+
+  } catch (std::ifstream::failure &) {
+    return header;
```
actually, this should never occur unless the file is completely destroyed, removed the try statement, so it will crash in that case

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  return result;
+}
+
+namespace {
```
merged and put at the top

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+  //             << ts2  << " "
+  //             << ts3  << " "
+  //             << header.timestamp  << " "
+  //             << std::endl;
```
I removed it, while it might be useful for testing, I do not want to have it in the release.
Since if you set the output to debug and run it on a normal sized datafile, you will get prints to the console for hours.

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+      return;
+    }
+    eventAccumulator.triggerEvents.push_back(event);
+  } break;
```
both are now functions returning small structs

thomueller committed:
Merge branch 'mantidproject:main' into 22871_read_DNS_psd_event_data

thomueller committed:
fix check for dataID, was asignement

thomueller committed:
add stats.h for bitstream, and fix jenkins warnings

thomueller committed:
fix renaming in test

thomueller commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,525 @@
+} // namespace
+
+std::vector<uint8_t> LoadDNSEvent::parse_Header(FileByteStream &file) {
+  // using Boyer-Moore String Search:
+  static const auto skipTable = buildSkipTable(header_sep);
```
The skip tables build by the boost algorithm are private.

DavidFair commented:
So I've double checked with Martyn and we would be happy to take this into 6.3 still for a number of reasons:

- This has been open for a while
- You couldn't do development on it through no fault of your own
- All the changes are isolated to a loader / associated files. So if it did break, it wouldn't affect any other workflows at all.

@thomueller Could you change the base branch (at the top of the PR) to `release-next` to target the next release?

DavidFair commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,479 @@
+    PARALLEL_END_INTERUPT_REGION
+    oversizedChanelIndexCounterA += oversizedChanelIndexCounter;
+    oversizedPosCounterA += oversizedPosCounter;
+    PARALLEL_CHECK_INTERUPT_REGION
```
`CHECK_INTERUPT_REGION` neds to move after the loop (swap it with the line below)

A nice example I refer back to often can be found here: https://developer.mantidproject.org/MultiThreadingInAlgorithms.html

thomueller replied:
ah I understand

DavidFair commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,479 @@
+  }
+
+  // parse individual file chuncks:
+  PARALLEL_FOR_IF(true)
```
This can be replaced with `PARALLEL_FOR_NO_WSP_CHECK` as we're not handling workspaces

thomueller replied:
ok thanks

DavidFair commented on Framework/DataHandling/src/LoadDNSEvent.cpp:
```diff
@@ -0,0 +1,479 @@
+
+  // parse individual file chuncks:
+  PARALLEL_FOR_IF(true)
+  for (int i = 0; i < static_cast<int>(filechuncks.size()); ++i) {
```
Can we make `int i` into `size_t i` then all the static casts go away

thomueller replied:
oh jeah, that's nice

thomueller replied:
or, that is not working, it says the index variable for OpenMP must be a signed integer

DavidFair replied:
Ah yes I forgot MSVC has an old OpenMP standard :face_exhaling:

DavidFair reviewed (commented):
Some minor fix-ups, overall the code is looking much cleaner and easier to review. Great work on the changes so far :+1:

thomueller committed:
remove custom boyer moore and add new minor fixes

DavidFair commented:
@thomueller You've merge main in when we're now based on release next.

Could you fix this as follows:
```
# Check you don't have any uncommitted changes, if you do I'd recommend you stash them
git status
```

```
# Point back to pre-merge commit
git reset bee04ac 
# Clean up files from merge commit we've unstaged
git reset --h

# Merge release-next instead. This assumes upstream is this repo. 
# I notice your using a fork, so use "upstream" or your equivalent for the following
git fetch upstream
git merge upstream/release-next
# Check the history shows your last commit and the merge commit as the last two 
git log
```

Once that is done re-push your changes with
`git push --force-with-lease` 
This will re-write the correct history instead

If you don't have upstream or equivalent in your local repo you can add it by simply doing:
`git remote add upstream https://github.com/mantidproject/mantid.git`
Where `upstream` can be replaced with a name of your choice

thomueller committed:
Merge remote-tracking branch 'upstream/release-next' into 22871_read_DNS_psd_event_data

thomueller commented:
Ah sorry for this David, I just clicked the merge button on the web interface, not recognizing that we now have to fetch the next-release.
Thanks for the help.

DavidFair reviewed (approved):
I've tested this locally and it works for me

I'm sure there are minor bits we can still do, but perfect is the enemy of good and I'd rather see something shipped this release than next

LGTM, also for the Gatekeeper I've checked with @sf1919 and she's happy with letting this release note go straight into the corresponding file as the PR has been open for a while

thomueller commented:
Great thanks a lot for all the work you put into it.

martyngigg commented:
I was all ready to hit merge but I think something has gone wrong with rebasing for `release-next`. There are lots of file changes now that I wouldn't expect and I think they are changes that are in `main` so we don't want them in the release branch and this point.

I'm happy to sort this out and get the contribution merged.

martyngigg commented:
We determined it would be easier to take the state of the files at the branch tip and put them on a new branch this is done in #33301 so I will now close this. @DavidFair and I will work to get it merged.
